### PR TITLE
better gadget handling for UnitPreDamaged

### PR DESCRIPTION
**Pull request by KyleAnthonyShepherd**
Multiple gadgets are trying to handle UnitPreDamaged. Usually they deal with separate cases, but sometimes multiple gadgets are trying to influence the same damage event, leading to some unpredictable effects. 

There is now a caution message if retDamage or retImpulse are modified 2 or more times, to warn devs and contributors if a preexisting gadget is altering these values. 

Also a minor change to retImpulse. Return values now multiply by the current retImpulse instead of overwriting. Way too many gadgets were returning a impulseMult = 1, overwriting the previous gadget's changes to this value. 
The primary bug this fixes is a unitcannon exploit due to unit_combomb_full_damage.lua. It was trying to set impulse to zero, but other gadgets were overwriting this zero, giving full comblast impulse to units while only doing 400 damage.

### CODE REVIEW

**Comment by sprunk** [comment]
>  unpredictable effects

Effects are entirely predictable and depend on gadget layer.

**Comment by KyleAnthonyShepherd** [comment]
Nearly every gadget currently handling UnitPreDamaged is on layer = 0, and how the tiebreaking is done is unclear

**Comment by WatchTheFort** [comment]
Shouldn't the behaviour be consistent with other callins which can produce conflicting return values from multiple gadgets? If it's last one wins, then the one "true" gadget should be moved to a higher layer to ensure it processes last.

**Comment by sprunk** [comment]
> and how the tiebreaking is done is unclear

If you set layers probably you won't have to worry about tiebreaking.

**Comment by KyleAnthonyShepherd** [comment]
There are a couple of cases of the newdamage of one gadget being fed into another gadget to output a second newdamage.

Mostly comblast handling (mo_preventcombomb.lua and unit_combomb_full_damage.lua), and paralyze damage handling (unit_paralyze_damage_limit.lua and unit_paralyze_damage_multiplier.lua). 

So there may be no one "true" gadget for each damage instance. 

Explicit layering would help (especially as gadgets can apply a multiplicative factor, additive factor, or a clamp to damage and impulse values, so ordering matters a lot), but causes maintenance problems (see people setting layer = -math.huge), and requires a dev working on a gadget to be aware of the layer of all existing gadgets. 

But overall, the damage handling has been fine, nothing too fancy is being done right now. It was the impulseMult handling that was a problem, where the last gadget had all the say in that value. (which was probably unit_crashing_aircraft.lua on layer 1000, just returning a default impulseMult = 1 for everything except crashing aircraft)

**Comment by Beherith** [comment]
remindme 24h

**Comment by WatchTheFort** [comment]
@Beherith it's been 24h

File: luarules/gadgets.lua [hunk 1]
```diff
@@ -1595,10 +1611,21 @@ function gadgetHandler:UnitPreDamaged(unitID, unitDefID, unitTeam, damage, paral
...
 		end
 		if imp ~= nil then
-			retImpulse = imp
+			if retImpulse ~= 1.0 and imp ~= 1 and UnitPreDamaged_impulse_caution_count == 0 then
+			    -- to caution devs if multiple gadgets are trying to adjust impulse
+				Spring.Echo("GadgetHandler: caution, multiple gadgets influencing impulse")
+				UnitPreDamaged_impulse_caution_count = UnitPreDamaged_impulse_caution_count + 1
+			end
+			--assuming multiple gadgets influencing impulse should stack multiplicative
+			retImpulse = retImpulse*imp
```
Comment (sprunk) [inline luarules/gadgets.lua:1628]: Personally I'd keep just this single line out of this whole PR (but it's an excellent change).

**Comment by KyleAnthonyShepherd** [comment]
Removed the echo warnings. Put in code comments instead, as a note to future devs.

**Review by sprunk — APPROVED** [review]

**Pull request merged**